fem/transfer.hpp

```diff
@@ -256,21 +341,56 @@ class L2ProjectionGridTransfer : public GridTransfer
       /// conservative left-inverse prolongation operation. This functionality
       /// is also provided as an Operator by L2Prolongation.
       void ProlongateTranspose(const Vector& x, Vector& y) const override;
+
+      void DeviceProlongateTranspose(const Vector& x, Vector& y) const;
+
       void SetRelTol(real_t p_rtol_) override { } ///< No-op.
       void SetAbsTol(real_t p_atol_) override { } ///< No-op.
+
+      //friend class L2ProjectionGridTransfer;
    };
 
+protected:
+
+   friend class L2ProjectionL2Space;
+
+   //Class below must be public as we now have device code
+public:
+
    /** Projection operator between a H1 high-order finite element space on a
        coarse mesh, and a H1 low-order finite element space on a refined mesh
        (LOR). */
    class L2ProjectionH1Space : public L2Projection
    {
+      const bool use_device, verify_solution;
+
    public:
       L2ProjectionH1Space(const FiniteElementSpace &fes_ho_,
-                          const FiniteElementSpace &fes_lor_);
+                          const FiniteElementSpace &fes_lor_,
+                          const bool use_device_,
+                          const bool verify_solution_,
+                          Coefficient *coeff_,
+                          MemoryType d_mt_ = MemoryType::HOST);
 #ifdef MFEM_USE_MPI
       L2ProjectionH1Space(const ParFiniteElementSpace &pfes_ho_,
-                          const ParFiniteElementSpace &pfes_lor_);
+                          const ParFiniteElementSpace &pfes_lor_,
+                          const bool use_device_,
+                          const bool verify_solution_,
+                          Coefficient *coeff_,
+                          MemoryType d_mt_ = MemoryType::HOST);
+#endif
+      /// Same as above but assembles action of R through 4 parts:
+      ///   ( )  inv( lumped(M_L) ), which is a diagonal matrix (essentially a vector)
+      ///   ( )  ElementRestrictionOperator for LOR space
+      ///   ( )  mixed mass matrix M_{LH}
+      ///   ( )  ElementRestrictionOperator for HO space
+      void DeviceL2ProjectionH1Space(const FiniteElementSpace &fes_ho_,
+                                     const FiniteElementSpace &fes_lor_,
+                                     const Coefficient* coeff_);
+#ifdef MFEM_USE_MPI
+      void DeviceL2ProjectionH1Space(const ParFiniteElementSpace &pfes_ho_,
+                                     const ParFiniteElementSpace &pfes_lor_,
+                                     const Coefficient* coeff_);
 #endif
```

Comment:
The parameters of these methods are redundant -- they are already available as class members. To distinguish the parallel version from the serial (when the parameters are removed), just call the parallel one `ParDeviceL2ProjectionH1Space`.